File: 递归与分治/fib.c
//用迭代打印出前40位斐波那契数列数。

//方法一：
#include <stdio.h>

int main()
{
    int i;
    int a[40];

    a[0] = 0;
    a[1] = 1;
    printf("%d %d ", a[0], a[1]);

    for (i = 2; i < 40; i++)
    {
        a[i] = a[i - 1] + a[i - 2];
        printf("%d ", a[i]);
    }

    return 0;
}

//方法二
#include <stdio.h>

int Fib(int i)
{
    if (i < 2)
    {
        return i == 0 ? 0 : 1;
    }

    return Fib(i - 1) + Fib(i - 2);
}

int main()
{
    int i;
    for (i = 0; i < 40; i++)
    {
        printf("%d ", Fib(i));
    }

    return 0;
}

File: Python版本/递归与分治/fib.py

def fib(n):
    """迭代法实现斐波那契数列"""
    a, b = 0, 1
    print(b)

    for _ in range(n-1):
        a, b = b, a + b
        print(b)


def fib2(n):
    """递归实现斐波那契数列"""
    if n == 1:
        return 1
    if n == 2:
        return 1

    return fib2(n - 1) + fib2(n - 2)



if __name__ == "__main__":
    # fib(40)
    print(fib2(5))